warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/edge/ProtobufEdgeDataTypeHandler.java

```diff
@@ -558,11 +559,31 @@ public long process(KEYIN key, RawRecordContainer event, Multimap<String,Normali
         /*
          * Create Edge Values from Edge Definitions
          */
+        
+        Multimap<String,NormalizedContentInterface> mSource = null;
+        Multimap<String,NormalizedContentInterface> mSink = null;
+        
         for (EdgeDefinition edgeDef : edgeDefs) {
             arithmetic.clearMatchingGroups();
             Map<String,Set<String>> matchingGroups = new HashMap<>();
             String jexlPreconditions = null;
             
+            // don't bother evaluating preconditions if we know this event doesn't have the necessary fields for this edge
+            if (depthFirstList.containsKey(edgeDef.getSourceFieldName()) && depthFirstList.containsKey(edgeDef.getSinkFieldName())) {
+                mSource = depthFirstList.get(edgeDef.getSourceFieldName());
+                mSink = depthFirstList.get(edgeDef.getSinkFieldName());
+            } else {
+                continue;
+            }
+            
+            // bail if the event doesn't contain any values for the source or sink field
+            if (null == mSource || null == mSink) {
```

Comment:
So.... why check for containsKey above if this check would catch the case where one of the fields does not exist?

Reply:
Also I think CollectionUtils.isEmpty(collection) will take care of both == null and .isEmpty().  A little more succinct.

Reply:
This is the old code, just moved up so we're not needlessly checking the jexl evaluation on every record.  Good suggestions, I was just going to save this kind of cleanup for the refactor MR.

Reply:
Yep.  Planning to clean this kind of thing up in the refactor MR.